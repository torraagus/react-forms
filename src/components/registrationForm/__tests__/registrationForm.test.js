import React from "react";
import ReactDOM from "react-dom";

import { render, cleanup, fireEvent, act } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

import RegistrationForm from "../registrationForm.component";

beforeEach(cleanup);

test("render without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<RegistrationForm />, div);
});

test("matches snapshot", () => {
  const tree = renderer.create(<RegistrationForm />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("calls onSubmit with the username, email and password when submit is clicked", async () => {
  const { getByTestId } = render(<RegistrationForm />);
  const user = {
    username: "test",
    email: "test@gmail.com",
    password: "passtest",
  };

  await act(async () => {
    fireEvent.change(getByTestId("username"), {
      target: { value: user.username },
    });
  });
  expect(getByTestId("username").value).toBe(user.username);

  await act(async () => {
    fireEvent.change(getByTestId("email"), {
      target: { value: user.email },
    });
  });
  expect(getByTestId("email").value).toBe(user.email);

  await act(async () => {
    fireEvent.change(getByTestId("password"), {
      target: { value: user.password },
    });
  });
  expect(getByTestId("password").value).toBe(user.password);

  await act(async () => {
    fireEvent.submit(getByTestId("form"));
  });
});

test("shows error messages in each field when submit is clicked before complete the form", async () => {
  const { getByTestId } = render(<RegistrationForm />);

  await act(async () => {
    fireEvent.submit(getByTestId("form"));
  });

  expect(getByTestId("username").value).toBe("");
  expect(getByTestId("usernameErrors").textContent).toBe(
    "The username field is required"
  );

  expect(getByTestId("email").value).toBe("");
  expect(getByTestId("emailErrors").textContent).toBe(
    "The email field is required"
  );

  expect(getByTestId("password").value).toBe("");
  expect(getByTestId("passwordErrors").textContent).toBe(
    "The password field is required"
  );
});

test("shows error messages in each field when submit is clicked after complete the form", async () => {
  const { getByTestId } = render(<RegistrationForm />);

  await act(async () => {
    fireEvent.change(getByTestId("username"), {
      target: { value: "this is a username really really long" },
    });
  });
  expect(getByTestId("username").value).toBe(
    "this is a username really really long"
  );

  await act(async () => {
    fireEvent.change(getByTestId("email"), { target: { value: "test@" } });
  });
  expect(getByTestId("email").value).toBe("test@");

  await act(async () => {
    fireEvent.change(getByTestId("password"), { target: { value: "pass" } });
  });
  expect(getByTestId("password").value).toBe("pass");

  await act(async () => {
    fireEvent.submit(getByTestId("form"));
  });
  expect(getByTestId("usernameErrors").textContent).toBe(
    "The username max length is 30 characters"
  );
  expect(getByTestId("emailErrors").textContent).toBe(
    "The email format is wrong"
  );
  expect(getByTestId("passwordErrors").textContent).toBe(
    "The password must have at least 8 characters"
  );
});

test("show password when toggle password button is pressed", async () => {
  const { getByTestId } = render(<RegistrationForm/>);

  await act( async () => {
    fireEvent.click(getByTestId("togglePassword"));
  });

  expect(getByTestId("password").type).toBe("text");
});

test("hide password when toggle password button is pressed", async () => {
  const { getByTestId } = render(<RegistrationForm/>);

  //show password
  await act( async () => {
    fireEvent.click(getByTestId("togglePassword"));
  });

  //hide password
  await act( async () => {
    fireEvent.click(getByTestId("togglePassword"));
  });

  expect(getByTestId("password").type).toBe("password");
});
