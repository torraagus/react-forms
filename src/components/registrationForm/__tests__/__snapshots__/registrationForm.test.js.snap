// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="mt-5 container-fluid"
>
  <div
    className="row center-xs"
  >
    <div
      className=""
    >
      <h2>
        Registration
      </h2>
    </div>
  </div>
  <div
    className="row center-xs"
  >
    <div
      className=""
    >
      <p>
        Please, complete the form below to create your account
      </p>
    </div>
  </div>
  <div
    className="mt-3 row"
  >
    <div
      className="col-xl-offset-4 col-xl-4 col-md-offset-3 col-md-6 col-sm-offset-2 col-sm-8 col-xs-offset-1 col-xs-10"
    >
      <form
        className=""
        data-testid="form"
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="formBasicUserName"
          >
            Username
          </label>
          <input
            className="form-control"
            data-testid="username"
            id="formBasicUserName"
            name="username"
            placeholder="Enter username"
            type="text"
          />
          <small
            className="text-muted form-text"
            data-testid="usernameErrors"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="formBasicEmail"
          >
            Email address
          </label>
          <input
            className="form-control"
            data-testid="email"
            id="formBasicEmail"
            name="email"
            placeholder="Enter email"
            type="email"
          />
          <small
            className="text-muted form-text"
            data-testid="emailErrors"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="formBasicPassword"
          >
            Password
          </label>
          <input
            className="form-control"
            data-testid="password"
            id="formBasicPassword"
            name="password"
            placeholder="Enter password"
            type="password"
          />
          <small
            className="text-muted form-text"
            data-testid="passwordErrors"
          />
        </div>
        <button
          className="w-100 btn btn-primary btn-sm"
          disabled={false}
          type="submit"
        >
          Create account
        </button>
        <div
          className="row center-xs"
        >
          <div
            className=""
          >
            <br />
            <small>
              Al registrarte, aceptas las
               
              <a
                href="#"
              >
                Condiciones del Servicio
              </a>
               y la
               
              <a
                href="#"
              >
                Pol√≠tica de Privacidad.
              </a>
            </small>
          </div>
        </div>
      </form>
    </div>
  </div>
  <br />
  <div
    className="row center-xs"
  >
    <div
      className=""
    />
  </div>
</div>
`;
